sage/example-transition/src/transition/hero_jam.rs

```diff
@@ -0,0 +1,266 @@
+						}
+
+						asset.fatigue =
+							(asset.fatigue as i32).saturating_add(fatigue).clamp(0, 255) as u8;
+						asset.energy =
+							(asset.energy as i32).saturating_add(energy).clamp(0, 255) as u8;
+
+						asset.state_type = StateType::Work;
+						asset.state_sub_type = *work_type as u8;
+						asset.state_sub_value = *work_time as u8;
+						asset.state_change_block_number =
+							Self::get_block_time_for_action(work_time);
+
+						Ok(vec![TransitionOutput::Mutated(asset_ids[0], Asset::from(asset))])
+					},
+				}
+			},
+			_ => Ok(vec![]),
+		}
+	}
+
+	fn get_block_time_for_action(action_time: &ActionTime) -> BlockNumber {
+		let current_block = ChainHandler::get_current_block_number();
+		let block_time = action_time.get_block_time_from();
+
+		current_block.saturating_add(block_time.saturated_into())
+	}
+
+	fn get_resource_fatigue(state_type: StateType, state_sub_type: u8, action_time: u8) -> i32 {
+		match state_type {
+			StateType::None => 0,
+			StateType::Sleep => match SleepType::from(state_sub_type) {
+				SleepType::Normal => -((action_time as i32) * (action_time as i32) * 11),
+				_ => 0,
+			},
+			StateType::Work => (action_time as i32) * 8,
+		}
+	}
+
+	fn get_resource_energy(state_type: StateType, state_sub_type: u8, action_time: u8) -> i32 {
+		match state_type {
+			StateType::None => 0,
+			StateType::Sleep => match SleepType::from(state_sub_type) {
+				SleepType::Normal => -(action_time as i32),
+				_ => 0,
+			},
+			StateType::Work => match WorkType::from(state_sub_type) {
+				WorkType::Hunt => -((action_time as i32) * 6),
+				_ => 0,
+			},
+		}
+	}
+}
+
+impl<AccountId, BlockNumber, AssetHandler, ChainHandler> SageGameTransition
+	for HeroJamTransition<AccountId, BlockNumber, AssetHandler, ChainHandler>
+where
+	AccountId: Member + Codec,
+	BlockNumber: BlockNumberT,
+	AssetHandler: AssetManager<AccountId = AccountId, AssetId = AssetId, Asset = Asset<BlockNumber>>
+		+ AssetInspector<AccountId = AccountId, AssetId = AssetId, Asset = Asset<BlockNumber>>,
+	ChainHandler: ChainInspector<BlockNumber = BlockNumber>,
+{
+	type TransitionId = HeroAction;
+	type TransitionConfig = ();
+	type AccountId = AccountId;
+	type AssetId = AssetId;
+	type Asset = Asset<BlockNumber>;
+	type Extra = ();
+
+	fn verify_rule(
+		transition_id: &Self::TransitionId,
+		account_id: &Self::AccountId,
+		asset_ids: &[Self::AssetId],
+		_: &Self::Extra,
+	) -> Result<(), RuleError> {
+		match transition_id {
+			HeroAction::Create => {
+				ensure_asset_length(asset_ids, 0)?;
+				ensure_account_has_not_asset_of_type::<_, _, AssetHandler>(
+					account_id,
+					AssetType::Hero,
+				)
+			},
+			HeroAction::Sleep(_) | HeroAction::Work(_, _) => {
+				ensure_asset_length(asset_ids, 1)?;
+				ensure_owner_of::<_, _, AssetHandler>(asset_ids, account_id)?;
+				ensure_all_asset_type::<_, _, AssetHandler>(asset_ids, AssetType::Hero)?;
+				ensure_can_state_change::<_, _, AssetHandler, ChainHandler>(&asset_ids[0])
+			},
```

Comment:
I just realized, in all these functions, we read the asset from storage redundantly. I guess, we want to read all the assets once at the beginning of the rule verification and then just run checks on them without reading again from storage.

Reply:
Actually, it is still redundant, because we read them in the transition execution too. Shall we switch to a paradigm like?

```rust
fn execute_transition(...) {
    let assets = get_assets()?;

    // maybe we have to pass the asset_ids too for certain checks
    let Self::verify_rule(&assets, ...)?;

    Self::do_transition(assets, ...)
}
```

What do you think? Tomorrow me optimization?